File: common/FireColumnModel.cpp
#include "FireColumnModel.hpp"
#include "utils.hpp"
#include "readcsv.hpp"
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <algorithm>
#include <filesystem>
#include <iostream>
#include <iomanip>

FireColumnModel::FireColumnModel() 
    : _min_latitude(0.0), _max_latitude(0.0), _min_longitude(0.0), _max_longitude(0.0),
      _bounds_initialized(false) {
    _datetime_range.resize(2);
}

FireColumnModel::~FireColumnModel() = default;

void FireColumnModel::readFromDirectory(const std::string& directoryPath, const std::vector<std::string>& allowedSubdirs) {
    auto csvFiles = getCSVFiles(directoryPath, allowedSubdirs);
    
    if (csvFiles.empty()) {
        std::cout << "[FireColumnModel] No CSV files found in: " << directoryPath << std::endl;
        return;
    }
    
    if (!allowedSubdirs.empty()) {
        std::cout << "[FireColumnModel] Processing " << csvFiles.size() << " CSV files from " 
                  << allowedSubdirs.size() << " partitioned subdirectories..." << std::endl;
    } else {
        std::cout << "[FireColumnModel] Processing " << csvFiles.size() << " CSV files..." << std::endl;
    }
    
    for (const auto& file : csvFiles) {
        try {
            readFromCSV(file);
        } catch (const std::exception& e) {
            std::cerr << "[FireColumnModel] Error processing " << file << ": " << e.what() << std::endl;
        }
    }
    
    std::cout << "[FireColumnModel] Loaded " << measurementCount() << " measurements from " 
              << siteCount() << " sites" << std::endl;
}

void FireColumnModel::readFromCSV(const std::string& filename) {
    CSVReader reader(filename);
    
    try {
        reader.open();
    } catch (const std::exception& e) {
        throw std::runtime_error("Failed to open CSV file " + filename + ": " + e.what());
    }
    
    std::vector<std::string> row;
    bool headerSkipped = false;
    
    while (reader.readRow(row)) {
        if (!headerSkipped) {
            headerSkipped = true;
            continue;
        }
        
        if (row.size() < 13) {
            continue;
        }
        
        try {
            double latitude = std::stod(row[0]);
            double longitude = std::stod(row[1]);
            std::string datetime = row[2];
            std::string parameter = row[3];
            double concentration = std::stod(row[4]);
            std::string unit = row[5];
            double raw_concentration = std::stod(row[6]);
            int aqi = std::stoi(row[7]);
            int category = std::stoi(row[8]);
            std::string site_name = row[9];
            std::string agency_name = row[10];
            std::string aqs_code = row[11];
            std::string full_aqs_code = row[12];
            
            insertMeasurement(latitude, longitude, datetime, parameter, concentration,
                            unit, raw_concentration, aqi, category, site_name,
                            agency_name, aqs_code, full_aqs_code);
            
        } catch (const std::exception& e) {
            continue;
        }
    }
    
    reader.close();
}

void FireColumnModel::insertMeasurement(double latitude, double longitude, const std::string& datetime,
                                       const std::string& parameter, double concentration, const std::string& unit,
                                       double raw_concentration, int aqi, int category,
                                       const std::string& site_name, const std::string& agency_name,
                                       const std::string& aqs_code, const std::string& full_aqs_code) {
    _latitudes.push_back(latitude);
    _longitudes.push_back(longitude);
    _datetimes.push_back(datetime);
    _parameters.push_back(parameter);
    _concentrations.push_back(concentration);
    _units.push_back(unit);
    _raw_concentrations.push_back(raw_concentration);
    _aqis.push_back(aqi);
    _categories.push_back(category);
    _site_names.push_back(site_name);
    _agency_names.push_back(agency_name);
    _aqs_codes.push_back(aqs_code);
    _full_aqs_codes.push_back(full_aqs_code);
    
    std::size_t newIndex = _latitudes.size() - 1;
    updateIndices(newIndex);
    updateGeographicBounds(latitude, longitude);
    updateDatetimeRange(datetime);
    
    _unique_sites.insert(site_name);
    _unique_parameters.insert(parameter);
    _unique_agencies.insert(agency_name);
}

std::vector<std::size_t> FireColumnModel::getIndicesBySite(const std::string& siteName) const {
    auto it = _site_indices.find(siteName);
    return (it != _site_indices.end()) ? it->second : std::vector<std::size_t>{};
}

std::vector<std::size_t> FireColumnModel::getIndicesByParameter(const std::string& parameter) const {
    auto it = _parameter_indices.find(parameter);
    return (it != _parameter_indices.end()) ? it->second : std::vector<std::size_t>{};
}

std::vector<std::size_t> FireColumnModel::getIndicesByAqsCode(const std::string& aqsCode) const {
    auto it = _aqs_indices.find(aqsCode);
    return (it != _aqs_indices.end()) ? it->second : std::vector<std::size_t>{};
}

void FireColumnModel::getGeographicBounds(double& min_lat, double& max_lat, 
                                         double& min_lon, double& max_lon) const {
    if (_bounds_initialized) {
        min_lat = _min_latitude;
        max_lat = _max_latitude;
        min_lon = _min_longitude;
        max_lon = _max_longitude;
    } else {
        min_lat = max_lat = min_lon = max_lon = 0.0;
    }
}

void FireColumnModel::updateIndices(std::size_t index) {
    if (index >= _site_names.size()) return;
    
    _site_indices[_site_names[index]].push_back(index);
    _parameter_indices[_parameters[index]].push_back(index);
    _aqs_indices[_aqs_codes[index]].push_back(index);
}

void FireColumnModel::updateGeographicBounds(double latitude, double longitude) {
    if (!_bounds_initialized) {
        _min_latitude = _max_latitude = latitude;
        _min_longitude = _max_longitude = longitude;
        _bounds_initialized = true;
    } else {
        _min_latitude = std::min(_min_latitude, latitude);
        _max_latitude = std::max(_max_latitude, latitude);
        _min_longitude = std::min(_min_longitude, longitude);
        _max_longitude = std::max(_max_longitude, longitude);
    }
}

void FireColumnModel::updateDatetimeRange(const std::string& datetime) {
    if (_datetime_range[0].empty() || datetime < _datetime_range[0]) {
        _datetime_range[0] = datetime;
    }
    if (_datetime_range[1].empty() || datetime > _datetime_range[1]) {
        _datetime_range[1] = datetime;
    }
}

std::vector<std::string> FireColumnModel::getCSVFiles(const std::string& directoryPath, const std::vector<std::string>& allowedSubdirs) const {
    std::vector<std::string> csvFiles;
    
    try {
        for (const auto& entry : std::filesystem::recursive_directory_iterator(directoryPath)) {
            if (entry.is_regular_file()) {
                const std::string filename = entry.path().string();
                
                // If we have partition restrictions, check if file is in allowed subdirectory
                if (!allowedSubdirs.empty()) {
                    bool allowed = false;
                    for (const auto& subdir : allowedSubdirs) {
                        // Build subdirectory path, handling trailing slashes
                        std::string subdirPath = directoryPath;
                        if (!subdirPath.empty() && subdirPath.back() != '/') {
                            subdirPath += "/";
                        }
                        subdirPath += subdir;
                        
                        if (filename.find(subdirPath) != std::string::npos) {
                            allowed = true;
                            break;
                        }
                    }
                    if (!allowed) continue;  // Skip this file
                }
                
                if (filename.size() >= 4 && 
                    filename.substr(filename.size() - 4) == ".csv") {
                    csvFiles.push_back(filename);
                }
            }
        }
    } catch (const std::filesystem::filesystem_error& e) {
        std::cerr << "[FireColumnModel] Error accessing directory " << directoryPath << ": " << e.what() << std::endl;
    }
    
    std::sort(csvFiles.begin(), csvFiles.end());
    return csvFiles;
}

File: common/fire_column_model.py
"""
FireColumnModel - Python Implementation
Column-oriented fire air quality data model for efficient querying
"""

import csv
import os
from typing import List, Dict, Set, Tuple, Optional
from collections import defaultdict


class FireColumnModel:
    """
    Column-oriented storage for fire air quality measurements.
    Provides efficient indexing and querying capabilities.
    """
    
    def __init__(self):
        # Columnar storage - parallel arrays
        self.latitudes: List[float] = []
        self.longitudes: List[float] = []
        self.datetimes: List[str] = []
        self.parameters: List[str] = []
        self.concentrations: List[float] = []
        self.units: List[str] = []
        self.raw_concentrations: List[float] = []
        self.aqis: List[int] = []
        self.categories: List[int] = []
        self.site_names: List[str] = []
        self.agency_names: List[str] = []
        self.aqs_codes: List[str] = []
        self.full_aqs_codes: List[str] = []
        
        # Index structures for fast lookups
        self._site_indices: Dict[str, List[int]] = defaultdict(list)
        self._parameter_indices: Dict[str, List[int]] = defaultdict(list)
        self._aqs_indices: Dict[str, List[int]] = defaultdict(list)
        
        # Metadata tracking
        self._unique_sites: Set[str] = set()
        self._unique_parameters: Set[str] = set()
        self._unique_agencies: Set[str] = set()
        self._datetime_range: List[str] = ["", ""]  # [min, max]
        
        # Geographic bounds
        self._min_latitude: Optional[float] = None
        self._max_latitude: Optional[float] = None
        self._min_longitude: Optional[float] = None
        self._max_longitude: Optional[float] = None
    
    def read_from_directory(self, directory_path: str, allowed_subdirs: List[str] = None) -> None:
        """
        Load all CSV files from a directory (recursively).
        
        Args:
            directory_path: Path to directory containing CSV files
            allowed_subdirs: Optional list of subdirectory names to load (for partitioning)
        """
        csv_files = self._get_csv_files(directory_path, allowed_subdirs)
        
        if not csv_files:
            print(f"[FireColumnModel] No CSV files found in: {directory_path}")
            return
        
        if allowed_subdirs:
            print(f"[FireColumnModel] Processing {len(csv_files)} CSV files from {len(allowed_subdirs)} subdirectories...")
        else:
            print(f"[FireColumnModel] Processing {len(csv_files)} CSV files from {directory_path}...")
        
        for csv_file in csv_files:
            try:
                self.read_from_csv(csv_file)
            except Exception as e:
                print(f"[FireColumnModel] Error processing {csv_file}: {e}")
        
        print(f"[FireColumnModel] Loaded {self.measurement_count()} measurements from {self.site_count()} sites")
    
    def read_from_csv(self, filename: str) -> None:
        """
        Load measurements from a single CSV file.
        
        Args:
            filename: Path to CSV file
        
        Expected CSV format (no header):
        latitude,longitude,datetime,parameter,concentration,unit,raw_concentration,aqi,category,site_name,agency_name,aqs_code,full_aqs_code
        """
        with open(filename, 'r', encoding='utf-8') as f:
            reader = csv.reader(f)
            
            for row in reader:
                if len(row) < 13:
                    continue  # Skip incomplete rows
                
                try:
                    # Parse and validate data
                    latitude = float(row[0].strip('"'))
                    longitude = float(row[1].strip('"'))
                    datetime = row[2].strip('"')
                    parameter = row[3].strip('"')
                    concentration = float(row[4].strip('"'))
                    unit = row[5].strip('"')
                    raw_concentration = float(row[6].strip('"'))
                    aqi = int(row[7].strip('"'))
                    category = int(row[8].strip('"'))
                    site_name = row[9].strip('"')
                    agency_name = row[10].strip('"')
                    aqs_code = row[11].strip('"')
                    full_aqs_code = row[12].strip('"')
                    
                    # Insert measurement
                    self.insert_measurement(
                        latitude, longitude, datetime, parameter,
                        concentration, unit, raw_concentration,
                        aqi, category, site_name, agency_name,
                        aqs_code, full_aqs_code
                    )
                    
                except (ValueError, IndexError) as e:
                    # Skip rows with invalid data
                    continue
    
    def insert_measurement(
        self,
        latitude: float,
        longitude: float,
        datetime: str,
        parameter: str,
        concentration: float,
        unit: str,
        raw_concentration: float,
        aqi: int,
        category: int,
        site_name: str,
        agency_name: str,
        aqs_code: str,
        full_aqs_code: str
    ) -> None:
        """Insert a single measurement into the columnar storage."""
        # Append to columns
        self.latitudes.append(latitude)
        self.longitudes.append(longitude)
        self.datetimes.append(datetime)
        self.parameters.append(parameter)
        self.concentrations.append(concentration)
        self.units.append(unit)
        self.raw_concentrations.append(raw_concentration)
        self.aqis.append(aqi)
        self.categories.append(category)
        self.site_names.append(site_name)
        self.agency_names.append(agency_name)
        self.aqs_codes.append(aqs_code)
        self.full_aqs_codes.append(full_aqs_code)
        
        # Update indices
        new_index = len(self.latitudes) - 1
        self._update_indices(new_index)
        self._update_geographic_bounds(latitude, longitude)
        self._update_datetime_range(datetime)
        
        # Update metadata
        self._unique_sites.add(site_name)
        self._unique_parameters.add(parameter)
        self._unique_agencies.add(agency_name)
    
    def get_indices_by_site(self, site_name: str) -> List[int]:
        """Get all measurement indices for a specific site."""
        return self._site_indices.get(site_name, [])
    
    def get_indices_by_parameter(self, parameter: str) -> List[int]:
        """Get all measurement indices for a specific parameter (e.g., PM2.5)."""
        return self._parameter_indices.get(parameter, [])
    
    def get_indices_by_aqs_code(self, aqs_code: str) -> List[int]:
        """Get all measurement indices for a specific AQS code."""
        return self._aqs_indices.get(aqs_code, [])
    
    def measurement_count(self) -> int:
        """Get total number of measurements."""
        return len(self.latitudes)
    
    def site_count(self) -> int:
        """Get number of unique sites."""
        return len(self._unique_sites)
    
    def unique_sites(self) -> Set[str]:
        """Get set of unique site names."""
        return self._unique_sites
    
    def unique_parameters(self) -> Set[str]:
        """Get set of unique parameters."""
        return self._unique_parameters
    
    def unique_agencies(self) -> Set[str]:
        """Get set of unique agencies."""
        return self._unique_agencies
    
    def datetime_range(self) -> Tuple[str, str]:
        """Get datetime range (min, max)."""
        return (self._datetime_range[0], self._datetime_range[1])
    
    def geographic_bounds(self) -> Tuple[Optional[float], Optional[float], Optional[float], Optional[float]]:
        """Get geographic bounds (min_lat, max_lat, min_lon, max_lon)."""
        return (self._min_latitude, self._max_latitude, self._min_longitude, self._max_longitude)
    
    def _update_indices(self, index: int) -> None:
        """Update index structures for a new measurement."""
        if index >= len(self.site_names):
            return
        
        self._site_indices[self.site_names[index]].append(index)
        self._parameter_indices[self.parameters[index]].append(index)
        self._aqs_indices[self.aqs_codes[index]].append(index)
    
    def _update_geographic_bounds(self, latitude: float, longitude: float) -> None:
        """Update geographic bounds tracking."""
        if self._min_latitude is None:
            self._min_latitude = self._max_latitude = latitude
            self._min_longitude = self._max_longitude = longitude
        else:
            self._min_latitude = min(self._min_latitude, latitude)
            self._max_latitude = max(self._max_latitude, latitude)
            self._min_longitude = min(self._min_longitude, longitude)
            self._max_longitude = max(self._max_longitude, longitude)
    
    def _update_datetime_range(self, datetime: str) -> None:
        """Update datetime range tracking."""
        if not self._datetime_range[0] or datetime < self._datetime_range[0]:
            self._datetime_range[0] = datetime
        if not self._datetime_range[1] or datetime > self._datetime_range[1]:
            self._datetime_range[1] = datetime
    
    def _get_csv_files(self, directory_path: str, allowed_subdirs: List[str] = None) -> List[str]:
        """
        Recursively find all CSV files in directory.
        
        Args:
            directory_path: Base directory path
            allowed_subdirs: Optional list of allowed subdirectory names (for partitioning)
        """
        csv_files = []
        
        try:
            for root, dirs, files in os.walk(directory_path):
                # If we have partition restrictions, check if current directory is allowed
                if allowed_subdirs:
                    # Get the immediate subdirectory name relative to directory_path
                    rel_path = os.path.relpath(root, directory_path)
                    # Check if this directory or its parent is in allowed list
                    dir_name = rel_path.split(os.sep)[0]
                    if dir_name != '.' and dir_name not in allowed_subdirs:
                        continue  # Skip this directory
                
                for file in files:
                    if file.endswith('.csv'):
                        csv_files.append(os.path.join(root, file))
        except Exception as e:
            print(f"[FireColumnModel] Error accessing directory {directory_path}: {e}")
        
        csv_files.sort()
        return csv_files

